### ISSUE TITLE
Remove mpi dependence from converters/*.py

### ISSUE THREAD

**Issue by shinaoka** [TRIQS/dft_tools#162]
### Summary

I propose to remove the (seemingly) unnecessary dependence of converters/*.py on MPI.

### Motivation
Importing mpi4py is prohibited on frontend nodes of some HPC clusters.
The mpi dependence prevents users from using these converters on such machines.

### Implementation
I think we can remove the MPI dependence if these converters are never run in parallel.
You can find some implementation based TRIQS 1.4 here.
https://github.com/issp-center-dev/DCore/tree/master/python/converters

Alternatively, we can check `'triqs.utility.mpi' in sys.modules` at runtime and
and import the mpi module in these converters only if true.

**Comment by the-hampel** [comment]
Hi,
I agree. I think it is reasonable to assume that all converters are run only by the master rank, if run with mpi at all, as they involve disk I/O which should only be done once. Indeed I am running the converters from an MPI environment within charge self-consistent DFT+DMFT calculations, but I only call the converters on the master rank. I think your PR #163 looks good as a workaround. Can anyone imagine a reason why the converters need MPI, any parallelization necessary for large amounts of k-points maybe? I can not find right now any use of MPI, except of mpi.report() which can be simply avoided.

I will discuss briefly tomorrow if someone as good reasons against it. I will merge your PR then as it has this nice backward compatibility layer. In a long run, I can also remove the use of MPI in all converters completely.

**Comment by aichhorn** [comment]
Thanks Hiroshi! This looks good to me. I thought a bit about mpi for the converters now, but I don't see a case where it would be necessary. There is nothing time critical in this part of the code.
From my side we can merge this.

**Comment by shinaoka** [comment]
Thank you for reviewing the PR!

**Comment by the-hampel** [comment]
Just an update: I am discussing with @Wentzell at the moment about the issue. In principle we agree, however we are not sure how "future" friendly an extra layer on top of the triqs.utility.mpi layer is, which itself is already a layer above mpi4py. Meaning, that if in future someone wants to use other MPI functions in the converters, they also need to be implemented in the extra layer. Right now we are checking if there are other options for solving the problem.

**Comment by the-hampel** [comment]
update 2: we think we have a more global solution to the problem. We will try to take your idea of a "dummy" mpi class directly into triqs. We have the idea that we can check when importing `triqs.utility.mpi` if a mpi environment is loaded by checking via `os.environ.get()` if MPI environment variables are set. If not, then we will basically do what you proposed. I will work on that the coming days. Do you think that could be also a possible alternative? Then every triqs part can be used with and without mpi, even if `triqs.utility.mpi` is imported.

**Comment by shinaoka** [comment]
Thank you for the updates.
Yes, introducing a dummy mpi class into TRIQS seems to be the best solution!

**Comment by shinaoka** [comment]
For your information, `from triqs_tprf.wannier90 import parse_hopping_from_wannier90_hr_dat` causes the same problem.

**Comment by the-hampel** [comment]
Just a quick update: I started working on the implementation and for the python layer I implemented some preliminary fixes that work quite well. You can check it out on my fork https://github.com/the-hampel/triqs/tree/mpi_simpl . I will continue working on it next week. Just to let you know, that I did not forgot about it. At the moment the C++ end is not yet fixed. 

Best,
Alex

**Comment by shinaoka** [comment]
Thank you for the update!
FYI, I have found similar issues in triqs.lattice module.

```
> from triqs.lattice.tight_binding import *
Starting run with 1 MPI rank(s) at : 2021-02-22 13:48:49.435287
> 
```

**Comment by the-hampel** [comment]
Dear @shinaoka,
@Wentzell and I have now implemented a no MPI mode in triqs. See PR: https://github.com/TRIQS/triqs/pull/809, https://github.com/TRIQS/mpi/pull/6, and https://github.com/TRIQS/nda/pull/13 for details. Right now this is due to technical reasons on the stat branch of triqs, but will be soonish merged to unstable. You can try the feature by using the stat branch of triqs. 

Our solution is now a bit more involved than just changing the python layer. The feature works as follows: Whenever in triqs MPI is invoked, no matter if in python or c++, we make a check if triqs is currently running via `mpirun` by checking the env variables for MPI related variables (which should work for all current supported MPI versions: OpenMPI, MPICH, cray, and Intel). If triqs was called via mpirun, MPI will be used as usual, but when triqs was called without mpirun, it will avoid any mpi calls, and not even make an mpi init. For this mode, dummy versions of the mpi functions are implemented. This should solve the problems you described. If there are any problems let us know. 

I tested the feature extensively for triqs, cthyb, and dfttools, and found no problems. 

Best,
Alex

**Comment by shinaoka** [comment]
@the-hampel 
Thank you for your effort!
These changes will be very useful.
I will test the new code next week.

**Comment by shinaoka** [comment]
I somehow failed to build the source files in the stat branch.
How soon will the stat branch be merged into the unstable branch? 

```
-- Found HDF5: /usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5.so;/usr/lib/x86_64-linux-gnu/libpthread.so;/usr/lib/x86_64-linux-gnu/libsz.so;/usr/lib/x86_64-linux-gnu/libz.so;/usr/lib/x86_64-linux-gnu/libdl.so;/usr/lib/x86_64-linux-gnu/libm.so (found version "1.10.4") found components: C HL
-- Adding cpp2py Python module storable
CMake Error at /home/shinaoka/build/triqs/deps/h5_src/python/h5/CMakeLists.txt:15 (install):
  install DIRECTORY given no DESTINATION!
```

**Comment by Wentzell** [comment]
Dear @shinaoka,

The stat branch has been merged into unstable last week.
Can you please try to build TRIQS/unstable instead?

Best,

Nils

**Comment by shinaoka** [comment]
@Wentzell 
For the unstable branch, the h5 library has been configured normally.
But, I see some compilations errors with gcc-9.3.0.
Do you have any idea what's wrong?

[CMakeCache.txt](https://github.com/TRIQS/dft_tools/files/6582027/CMakeCache.txt)
[compilation_log.txt](https://github.com/TRIQS/dft_tools/files/6582028/compilation_log.txt)

```
633 /home/shinaoka/build/triqs/deps/nda_src/c++/nda/./group_indices.hpp: In instantiation of ‘auto nda::group_indices_layout(const nda::idx_map<R, SE, SO, LP>&, IdxGrps ...) [with int Rank = 4; lo    ng unsigned int StaticExtents = 0; long unsigned int StrideOrder = 12816; nda::layout_prop_e LayoutProp = nda::layout_prop_e::contiguous; IdxGrps = {nda::idx_group_t<0, 1>, nda::idx_group_t<2,     3>}]’:
634 /home/shinaoka/build/triqs/deps/nda_src/c++/nda/./layout_transforms.hpp:137:56:   required from ‘auto nda::group_indices_view(nda::basic_array_view<ValueType, Rank, Layout, Algebra, AccessorPo    licy, OwningPolicy>, IntSequences ...) [with T = int; int R = 4; L = nda::C_layout; char Algebra = 'A'; AccessorPolicy = nda::default_accessor; OwningPolicy = nda::borrowed; IntSequences = {nda    ::idx_group_t<0, 1>, nda::idx_group_t<2, 3>}]’
635 /home/shinaoka/build/triqs/deps/nda_src/test/c++/nda_functions.cpp:88:74:   required from here
636 /home/shinaoka/build/triqs/deps/nda_src/c++/nda/./group_indices.hpp:149:57: error: non-constant condition for static assertion
637   149 |     static_assert(details::is_partition_of_indices<Rank>(IdxGrps::as_std_array...),
638       |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by Wentzell** [comment]
Dear @shinaoka,

Could you try compiling with gcc version 10 or above instead?
We are gradually starting our move to C++20 for our unstable branches,
and the error above is related to this.

Best,

Nils

**Comment by the-hampel** [comment]
Hi, 

I can confirm that the unstable branch can no longer be build with something like gcc/7.4.0. gcc/10.x works. I can reproduce the error that @shinaoka sees:

```
[ 34%] Building CXX object deps/nda/test/c++/CMakeFiles/nda_functions.dir/nda_functions.cpp.o                                                                                                 
In file included from /dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/c++/nda/./layout_transforms.hpp:17:0,                                                                           
                 from /dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/c++/nda/basic_array.hpp:363,                                                                                    
                 from /dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/c++/nda/nda.hpp:28,                                                                                             
                 from /dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/test/c++/test_common.hpp:25,                                                                                    
                 from /dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/test/c++/nda_functions.cpp:15:                                                                                  
/dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/c++/nda/./group_indices.hpp: In instantiation of 'auto nda::group_indices_layout(const nda::idx_map<R, SE, SO, LP>&, IdxGrps ...) [with int Rank = 4; long unsigned int StaticExtents = 0; long unsigned int StrideOrder = 12816; nda::layout_prop_e LayoutProp = (nda::layout_prop_e)3; IdxGrps = {nda::idx_group_t<0, 1>, nda::idx_
group_t<2, 3>}]':                                                                              
/dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/c++/nda/./layout_transforms.hpp:137:56:   required from 'auto nda::group_indices_view(nda::basic_array_view<ValueType, Rank, Layout, Algebra, AccessorPolicy, OwningPolicy>, IntSequences ...) [with T = int; int R = 4; L = nda::C_layout; char Algebra = 'A'; AccessorPolicy = nda::default_accessor; OwningPolicy = nda::borrowed;
 IntSequences = {nda::idx_group_t<0, 1>, nda::idx_group_t<2, 3>}]'                             
/dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/test/c++/nda_functions.cpp:88:74:   required from here                                                                                
/dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/c++/nda/./group_indices.hpp:149:5: error: non-constant condition for static assertion                                                 
     static_assert(details::is_partition_of_indices<Rank>(IdxGrps::as_std_array...),           
     ^~~~~~~~~~~~~                                                                             
/dev/shm/triqs3_unstable_build/triqs.build/deps/nda_src/c++/nda/./group_indices.hpp:149:57: error: 'constexpr bool nda::details::is_partition_of_indices(const std::array<int, Rs>& ...) [with long unsigned int R = 4; long unsigned int ...Rs = {2, 2}]' called in a constant expression
     static_assert(details::is_partition_of_indices<Rank>(IdxGrps::as_std_array...),   
```

- Alex

**Comment by shinaoka** [comment]
@Wentzell @the-hampel 
Thank you for your help.
The unstable branch has been built successfully with g++ 10.2.0.
The new code works well!

<img width="426" alt="image" src="https://user-images.githubusercontent.com/15949860/120549677-dc63cd80-c42e-11eb-809e-ba11d6c5814d.png">

**Comment by the-hampel** [comment]
Great! That is how it supposed to work now, giving a warning if started in serial mode. :-) Took quite some time, but finally. Let us know if there are any problems.

**Comment by shinaoka** [comment]
Thank you for your effort again. I close this issue.
